Update raw-window-handle to 0.6, core-graphics to 0.23, and cocoa to 0.25

Nopey opened the pull request:
These are the latest versions of each crate.

Nopey committed:
Update raw-window-handle to 0.6

Nopey committed:
Update core-graphics to 0.23, cocoa to 0.25

Nopey committed:
Add raw-window-handle features for 0.5 and 0.6

mrobinson commented on surfman/src/platform/generic/multi/connection.rs:
```diff
@@ -209,10 +222,27 @@ where
         &self,
-        raw_handle: raw_window_handle::RawWindowHandle,
+        raw_handle: rwh_06::RawWindowHandle,
         size: Size2D<i32>,
```
Hrm. This is a bit odd. If you are using the same type names, why would you need to duplicate this function based on the version of the library used? Note that in `winit` for raw-window-handle-0.6 it doesn't use `RawWindowHandle`, but `WindowHandle` and friends (https://docs.rs/winit/latest/winit/window/struct.Window.html#method.window_handle)

Nopey replied:
`winit` is using raw_window_handle's traits like HasRawWindowHandle, and implementing both.
raw_window_handle does not provide traits for our usecase, where we want to consume either a rwh 0.5 or 0.6

### two ways of hacking method overloading into rust

First approach to getting there, where each function gets two traits to implement overloading.
<details><summary>idea 1: create two traits (v0.5 and v0.6) for each function we have currently, e.g. trait CreateNativeWidgetFromRawWindowHandle05</summary>

```rs
#[cfg(feature = "sm-raw-window-handle-05")]
trait CreateNativeWidgetFromRawWindowHandle05 {
    fn create_native_widget_from_raw_window_handle(&self, raw_handle: rwh_05::RawWindowHandle, size: Size2D<32>, ) -> NativeWidget<Def, Alt>;
}
#[cfg(feature = "sm-raw-window-handle-05")]
impl<Def, Alt> CreateNativeWidgetFromRawWindowHandle05 for Connection<Def, Alt>
where
    Def: DeviceInterface,
    Alt: DeviceInterface,
    Def::Connection: ConnectionInterface<Device = Def>,
    Alt::Connection: ConnectionInterface<Device = Alt>,
    {
    fn create_native_widget_from_raw_window_handle(&self, raw_handle: rwh_05::RawWindowHandle, size: Size2D<32>, ) -> NativeWidget<Def, Alt>{
      // implementation goes here
    }
}
```
This seems like an untenable amount of boilerplate to do for each method.
User gain: they no longer have to call the _05 version when using the old version of raw window handle.
User loss: they now have to include the traits to call the method.
</details>
I don't think this first approach is an improvement over the current MR.

If we try to introduce an abstraction over the rwh 05/06 types themselves.. we get idea 2:
<details><summary> idea 2: impl Into&lt;enum RawWindowHandle&gt;</summary>

```rs
mod surfman {
    // putting this "public" enum in a private module bypasses warn(private_bounds) without
    // exposing the enum.. surprised the compiler allows this, might be better to #[allow(private_bounds)] or publicize this enum?
    mod impl_detail {
        pub enum RawWindowHandle {
            #[cfg(feature = "sm-raw-window-handle-05")]
            Rwh05(u32), // of course, replace this u32 with the real raw_window_handle 0.5 ::RawWindowHandle type
            #[cfg(feature = "sm-raw-window-handle-06")]
            Rwh06(String),
        }
    }
    use impl_detail::RawWindowHandle;
    
    #[cfg(feature = "sm-raw-window-handle-05")]
    impl From<u32> for RawWindowHandle {
        fn from(rwh: u32) -> Self {
            RawWindowHandle::Rwh05(rwh)
        }
    }
    
    #[cfg(feature = "sm-raw-window-handle-06")]
    impl From<String> for RawWindowHandle {
        fn from(rwh: String) -> Self {
            RawWindowHandle::Rwh06(rwh)
        }
    }
    
    // the idea with this function is to optimize into a direct call to the correct function
    #[inline(always)]
    pub fn consume_raw_window_handle(rwh: impl Into<RawWindowHandle>) {
        match rwh.into() {
            #[cfg(feature = "sm-raw-window-handle-05")]
            RawWindowHandle::Rwh05(rwh) => consume_raw_window_handle_05(rwh),
            #[cfg(feature = "sm-raw-window-handle-06")]
            RawWindowHandle::Rwh06(rwh) => consume_raw_window_handle_06(rwh),
        }
    }
    
    #[cfg(feature = "sm-raw-window-handle-05")]
    fn consume_raw_window_handle_05(rwh: u32) {
        println!("RWH v0.5: {rwh}");
    }
    
    #[cfg(feature = "sm-raw-window-handle-06")]
    fn consume_raw_window_handle_06(rwh: String) {
        println!("RWH v0.6: {rwh}");
    }
}

fn main() {
    // here in surfman user-code, we just call the public interface
    // with whatever version of raw_window_handle we happen to use
    surfman::consume_raw_window_handle(String::new());
}
```
This approach has better user ergonomics, although the generated documentation for the consume_raw_window_handle function has a 'RawWindowHandle' that doesn't link anywhere.
</details>

Should I implement the idea 2 style impl Into enum? or is it better to leave the _05 variant of the methods as-is.

### WindowHandle API

I'll push a commit that makes the rwh_06 versions of these methods use the new WindowHandle API :+1:

Nopey replied:
Oh and I should answer the question "why would you need to duplicate this function based on the version of the library used?" directly:
rwh06::RawWindowHandle and rwh05::RawWindowHandle are not the same type. Also, Cargo features are additive, so a hypothetical "one function to support them both" would have to support both types at once.. See prior response's idea 2, where an enum was used to this effect.

Nopey replied:
Latest patchset uses `from_raw_display_handle` and similar for 0.5 api, `from_display_handle` for 0.6 api-- no more _05 suffixes.

mrobinson commented on surfman/src/connection.rs:
```diff
@@ -55,9 +55,15 @@ pub trait Connection: Sized {
+    #[cfg(feature = "sm-raw-window-handle-05")]
+    fn from_raw_display_handle_05(
+        raw_handle: rwh_05::RawDisplayHandle,
+    ) -> Result<Self, Error>;
```
Instead of using the deprecated `RawWindowHandle` and `RawDisplayHandle` when `raw-window-handle` 0.6 is enabled, let's switch to using `DisplayHandle` and `WindowHandle`. That way instead of pairs of functions like:

 - `from_raw_display_handle_05`
 - `from_raw_display_handle`

We could have:

 - `from_display_handle`
 - `from_raw_display_handle`

Nopey replied:
With rwh 0.6, the non-raw API's central idea is ensuring that whatever entity (e.g. a Windows Window) backs the raw window handle (e.g. HWND) isn't destroyed until the reference to the handle is dropped.

Should I be using something like: `has_display_handle: Box<dyn HasDisplayHandle + 'static>` or `struct WidgetThatsCreatedFromDisplayHandle<'a>{ display_handle: DisplayHandle<'a> }` ?

I'm not familiar with GATs, but we might even be able to make each of the associated types on the Connection trait generic over the HasDisplayHandle and HasWindowHandle implementations?

mrobinson replied:
This seems like a larger refactor. For now can we simply create and consume these handles in the moment and then figure out how to keep the display primitives alive in a separate change?

Nopey replied:
Alright, that should do it.
- No more _05 suffixes
- sm-raw-w-indow-handle-06 feature exposes an interface that consumes DisplayHandles and returns surfman objects that contain their raw platform-specific handle.
Thanks for your help!

mrobinson reviewed (changes requested):
Sincere apologies for the delay in reviewing! This look good in theory with one significant change. Let's not use the deprecated types in the new version of `raw-window-handle` and that way we can avoid the oddly named `_05` functions in the public API (which aren't very future proof).

Nopey committed:
Fix rustdoc::bare_urls warning on Khronos link

Nopey committed:
Fix compilation errors in threads example

This example segfaults on my machine (even before this PR); despite
this, I think having it compile is valuable.

Nopey committed:
Update threads example to winit 0.29

Nopey committed:
Add rwh 0.6 support to threads example

mrobinson commented on surfman/examples/threads.rs:
```diff
@@ -86,9 +90,49 @@ static BACKGROUND_COLOR: [f32; 4] = [
+    connection
+}
+
+// rwh 0.6
```
```suggestion
```

mrobinson commented on surfman/examples/threads.rs:
```diff
@@ -86,9 +90,49 @@ static BACKGROUND_COLOR: [f32; 4] = [
+    native_widget
+}
+
+// rwh 0.6
```
```suggestion
```

mrobinson commented on surfman/src/connection.rs:
```diff
@@ -55,9 +55,15 @@ pub trait Connection: Sized {
+        raw_handle: rwh_05::RawDisplayHandle,
+    ) -> Result<Self, Error>;
+
+    /// Opens the display connection corresponding to the given raw display handle.
```
```suggestion
    /// Opens the display connection corresponding to the given `DisplayHandle`.
```

mrobinson commented on surfman/src/connection.rs:
```diff
@@ -68,10 +74,18 @@ pub trait Connection: Sized {
+        size: Size2D<i32>,
+    ) -> Result<Self::NativeWidget, Error>;
+
+    /// Create a native widget type from the given `raw_window_handle::RawWindowHandle`.
```
```suggestion
    /// Create a native widget type from the given `WindowHandle`.
```

mrobinson commented on surfman/src/platform/android/connection.rs:
```diff
@@ -98,9 +98,17 @@ impl Connection {
+        Ok(Connection)
+    }
+
+    /// Opens the display connection corresponding to the given raw display handle.
```
```suggestion
    /// Opens the display connection corresponding to the given `DisplayHandle`.
```

mrobinson commented on surfman/src/platform/android/connection.rs:
```diff
@@ -133,6 +141,24 @@ impl Connection {
         }
     }
+
+    /// Create a native widget type from the given `raw_window_handle::RawWindowHandle`.
```
```suggestion
    /// Create a native widget type from the given `WindowHandle`.
```

mrobinson commented on surfman/src/platform/generic/multi/connection.rs:
```diff
@@ -192,6 +192,19 @@ where
         }
     }
 
+    /// Opens the display connection corresponding to the given raw display handle.
```
```suggestion
    /// Opens the display connection corresponding to the given `DisplayHandle`.
```

mrobinson commented on surfman/src/platform/generic/multi/connection.rs:
```diff
@@ -224,6 +237,23 @@ where
         }
     }
+
+    /// Create a native widget type from the given `raw_window_handle::HasRawWindowHandle`.
```
```suggestion
    /// Create a native widget type from the given `WindowHandle`.
```

mrobinson commented on surfman/src/platform/generic/multi/connection.rs:
```diff
@@ -209,10 +222,10 @@ where
     }
 
     /// Create a native widget type from the given `raw_window_handle::HasRawWindowHandle`.
```
```suggestion
    /// Create a native widget type from the given `RawWindowHandle`.
```

mrobinson commented on surfman/src/platform/android/connection.rs:
```diff
@@ -117,14 +125,14 @@ impl Connection {
     }
 
     /// Create a native widget type from the given `raw_window_handle::RawWindowHandle`.
```
```suggestion
    /// Create a native widget type from the given `RawWindowHandle`.
```

mrobinson commented on surfman/src/connection.rs:
```diff
@@ -68,10 +74,18 @@ pub trait Connection: Sized {
     ) -> Self::NativeWidget;
 
     /// Create a native widget type from the given `raw_window_handle::RawWindowHandle`.
```
```suggestion
    /// Create a native widget type from the given `RawWindowHandle`.
```

mrobinson commented on surfman/src/connection.rs:
```diff
@@ -55,9 +55,15 @@ pub trait Connection: Sized {
     ) -> Result<Self::Device, Error>;
 
     /// Opens the display connection corresponding to the given raw display handle.
```
```suggestion
    /// Opens the display connection corresponding to the given `RawDisplayHandle`.
```

mrobinson commented on surfman/src/platform/macos/cgl/connection.rs:
```diff
@@ -100,13 +95,21 @@ impl Connection {
     }
 
     /// Opens the display connection corresponding to the given raw display handle.
```
```suggestion
    /// Opens the display connection corresponding to the given `RawDisplayHandle`.
```

mrobinson commented on surfman/src/platform/macos/cgl/connection.rs:
```diff
@@ -100,13 +95,21 @@ impl Connection {
         SystemConnection::from_raw_display_handle(raw_handle).map(Connection)
     }
 
+    /// Opens the display connection corresponding to the given raw display handle.
```
```suggestion
    /// Opens the display connection corresponding to the given `DisplayHandle`.
```

mrobinson commented on surfman/src/platform/macos/cgl/connection.rs:
```diff
@@ -117,14 +120,16 @@ impl Connection {
     }
 
     /// Create a native widget type from the given `raw_window_handle::RawWindowHandle`.
```
```suggestion
    /// Create a native widget type from the given `RawWindowHandle`.
```

mrobinson commented on surfman/src/platform/macos/cgl/connection.rs:
```diff
@@ -134,4 +139,25 @@ impl Connection {
         }
     }
+
+    /// Create a native widget type from the given `raw_window_handle::RawWindowHandle`.
```
```suggestion
    /// Create a native widget type from the given `WindowHandle`.
```

mrobinson commented on surfman/src/platform/macos/system/connection.rs:
```diff
@@ -119,9 +119,17 @@ impl Connection {
+        Connection::new()
+    }
+
+    /// Opens the display connection corresponding to the given raw display handle.
```
```suggestion
    /// Opens the display connection corresponding to the given `DisplayHandle`.
```

mrobinson commented on surfman/src/platform/macos/system/connection.rs:
```diff
@@ -119,9 +119,17 @@ impl Connection {
     }
 
     /// Opens the display connection corresponding to the given raw display handle.
```
```suggestion
    /// Opens the display connection corresponding to the given `RawDisplayHandle`.
```

mrobinson commented on surfman/src/platform/macos/system/connection.rs:
```diff
@@ -139,14 +147,14 @@ impl Connection {
     }
 
     /// Create a native widget type from the given `raw_window_handle::RawWindowHandle`.
```
```suggestion
    /// Create a native widget type from the given `RawWindowHandle`.
```

mrobinson commented on surfman/src/platform/macos/system/connection.rs:
```diff
@@ -156,6 +164,25 @@ impl Connection {
         }
     }
+
+    /// Create a native widget type from the given `raw_window_handle::RawWindowHandle`.
```
```suggestion
    /// Create a native widget type from the given `WindowHandle`.
```

mrobinson commented on surfman/src/platform/unix/generic/connection.rs:
```diff
@@ -135,9 +135,17 @@ impl Connection {
     }
 
     /// Opens the display connection corresponding to the given raw display handle.
```
```suggestion
    /// Opens the display connection corresponding to the given `RawDisplayHandle`.
```

mrobinson commented on surfman/src/platform/unix/generic/connection.rs:
```diff
@@ -135,9 +135,17 @@ impl Connection {
+        Err(Error::IncompatibleNativeWidget)
+    }
+
+    /// Opens the display connection corresponding to the given raw display handle.
```
```suggestion
    /// Opens the display connection corresponding to the given `DisplayHandle`.
```

mrobinson commented on surfman/src/platform/unix/generic/connection.rs:
```diff
@@ -152,11 +160,22 @@ impl Connection {
     }
 
     /// Create a native widget type from the given `raw_window_handle::RawWindowHandle`.
```
```suggestion
    /// Create a native widget type from the given `RawWindowHandle`.
```

mrobinson commented on surfman/src/platform/unix/generic/connection.rs:
```diff
@@ -152,11 +160,22 @@ impl Connection {
+        Err(Error::IncompatibleNativeWidget)
+    }
+
+    /// Create a native widget type from the given `raw_window_handle::RawWindowHandle`.
```
```suggestion
    /// Create a native widget type from the given `WindowHandle`.
```

mrobinson commented on surfman/src/platform/unix/wayland/connection.rs:
```diff
@@ -157,12 +157,12 @@ impl Connection {
     }
 
     /// Opens the display connection corresponding to the given raw display handle.
```
```suggestion
    /// Opens the display connection corresponding to the given `RawDisplayHandle`.
```

mrobinson commented on surfman/src/platform/unix/wayland/connection.rs:
```diff
@@ -173,6 +173,23 @@ impl Connection {
         }
     }
 
+    /// Opens the display connection corresponding to the given raw display handle.
```
```suggestion
    /// Opens the display connection corresponding to the given `DisplayHandle`.
```

mrobinson commented on surfman/src/platform/unix/wayland/connection.rs:
```diff
@@ -186,13 +203,13 @@ impl Connection {
     }
 
     /// Creates a native widget type from the given `raw_window_handle::HasRawWindowHandle`
```
```suggestion
    /// Creates a native widget type from the given `RawWindowHandle`
```

mrobinson commented on surfman/src/platform/unix/wayland/connection.rs:
```diff
@@ -204,6 +221,26 @@ impl Connection {
         })
     }
+
+    /// Creates a native widget type from the given `raw_window_handle::HasRawWindowHandle`
```
```suggestion
    /// Creates a native widget type from the given `WindowHandle`
```

mrobinson commented on surfman/src/platform/unix/x11/connection.rs:
```diff
@@ -216,13 +233,13 @@ impl Connection {
     }
 
     /// Create a native widget type from the given `raw_window_handle::HasRawWindowHandle`.
```
```suggestion
    /// Create a native widget type from the given `RawWindowHandle`.
```

mrobinson commented on surfman/src/platform/unix/x11/connection.rs:
```diff
@@ -231,6 +248,23 @@ impl Connection {
         }
     }
+
+    /// Create a native widget type from the given `raw_window_handle::HasRawWindowHandle`.
```
```suggestion
    /// Create a native widget type from the given `WindowHandle`.
```

mrobinson commented on surfman/src/platform/windows/angle/connection.rs:
```diff
@@ -129,9 +129,17 @@ impl Connection {
     }
 
     /// Opens the display connection corresponding to the given raw display handle.
```
```suggestion
    /// Opens the display connection corresponding to the given `RawDisplayHandle`.
```

mrobinson commented on surfman/src/platform/windows/angle/connection.rs:
```diff
@@ -129,9 +129,17 @@ impl Connection {
+        Connection::new()
+    }
+
+    /// Opens the display connection corresponding to the given raw display handle.
```
```suggestion
    /// Opens the display connection corresponding to the given `DisplayHandle`.
```

mrobinson commented on surfman/src/platform/windows/angle/connection.rs:
```diff
@@ -148,21 +156,38 @@ impl Connection {
     }
 
     /// Create a native widget type from the given `raw_window_handle::RawWindowHandle`.
```
```suggestion
    /// Create a native widget type from the given `RawWindowHandle`.
```

mrobinson commented on surfman/src/platform/windows/angle/connection.rs:
```diff
@@ -148,21 +156,38 @@ impl Connection {
         }
     }
+
+    /// Create a native widget type from the given `raw_window_handle::RawWindowHandle`.
```
```suggestion
    /// Create a native widget type from the given `WindowHandle`.
```

mrobinson commented on surfman/src/platform/windows/wgl/connection.rs:
```diff
@@ -100,9 +100,17 @@ impl Connection {
     }
 
     /// Opens the display connection corresponding to the given raw display handle.
```
```suggestion
    /// Opens the display connection corresponding to the given `RawDisplayHandle`.
```

mrobinson commented on surfman/src/platform/windows/wgl/connection.rs:
```diff
@@ -100,9 +100,17 @@ impl Connection {
+        Connection::new()
+    }
+
+    /// Opens the display connection corresponding to the given raw display handle.
```
```suggestion
    /// Opens the display connection corresponding to the given `DisplayHandle`.
```

mrobinson commented on surfman/src/platform/windows/wgl/connection.rs:
```diff
@@ -119,13 +127,13 @@ impl Connection {
     }
 
     /// Create a native widget type from the given `raw_window_handle::HasRawWindowHandle`.
```
```suggestion
    /// Create a native widget type from the given `RawWindowHandle`.
```

mrobinson commented on surfman/src/platform/windows/wgl/connection.rs:
```diff
@@ -134,6 +142,23 @@ impl Connection {
         }
     }
+
+    /// Create a native widget type from the given `raw_window_handle::WindowHandle`.
```
```suggestion
    /// Create a native widget type from the given `WindowHandle`.
```

mrobinson commented on surfman/src/platform/unix/x11/connection.rs:
```diff
@@ -204,6 +204,23 @@ impl Connection {
         Connection::from_x11_display(display, false)
     }
 
+    /// Opens the display connection corresponding to the given raw display handle.
```
```suggestion
    /// Opens the display connection corresponding to the given `DisplayHandle`.
```

mrobinson commented on surfman/src/platform/unix/x11/connection.rs:
```diff
@@ -188,13 +188,13 @@ impl Connection {
     }
 
     /// Opens the display connection corresponding to the given raw display handle.
```
```suggestion
    /// Opens the display connection corresponding to the given `RawDisplayHandle`.
```

mrobinson commented on surfman/src/platform/generic/multi/connection.rs:
```diff
@@ -180,9 +180,9 @@ where
     }
 
     /// Opens the display connection corresponding to the given raw display handle.
```
```suggestion
    /// Opens the display connection corresponding to the given `RawDisplayHandle.
```

mrobinson reviewed (approved):
Thanks. This looks great. I just have some suggestions around documentation consistency and correctness.

mrobinson committed:
Documentation fixes for consistency and correctness

mrobinson committed:
Update CI configuration for new features

mrobinson reviewed (changes requested):
Looks like there are some build failures on various platforms. These will need to be fixed before landing.

Nopey committed:
Fix Unix build

Nopey committed:
Fix Android build

Nopey committed:
Fix MacOS build

mrobinson committed:
Minor adjustments to documentation
